src/main/java/org/zalando/nakadi/service/EventTypeService.java

```diff
@@ -180,52 +181,50 @@ private void validateSchemaChange(final EventType eventType, final EventType exi
-                    .collect(Collectors.toList());
-            if (!absentFields.isEmpty()) {
-                throw new InvalidEventTypeException("partition_key_fields " + absentFields + " absent in schema");
+            final Schema schema = SchemaLoader.load(schemaAsJson);
+
+            if (eventType.getCategory() == EventCategory.BUSINESS && schema.definesProperty("#/metadata")) {
+                throw new InvalidEventTypeException("\"metadata\" property is reserved");
             }
+
+            validatePartitionKeys(Optional.of(schema), eventType);
+
         } catch (final JSONException e) {
             throw new InvalidEventTypeException("schema must be a valid json");
         } catch (final SchemaException e) {
             throw new InvalidEventTypeException("schema must be a valid json-schema");
         }
     }
 
-    private void validateSchema(final EventType eventType) throws InvalidEventTypeException {
-        try {
+    private void validatePartitionKeys(final Optional<Schema> schemaO, final EventType eventType)
+            throws InvalidEventTypeException, JSONException, SchemaException {
+        if (!featureToggleService.isFeatureEnabled(CHECK_PARTITIONS_KEYS)) {
+            return;
+        }
+        final Schema schema = schemaO.orElseGet(() -> {
             final JSONObject schemaAsJson = new JSONObject(eventType.getSchema().getSchema());
+            return SchemaLoader.load(schemaAsJson);
+        });
 
-            if (eventType.getCategory() == EventCategory.BUSINESS
-                    && hasReservedField(schemaAsJson, "metadata")) {
-                throw new InvalidEventTypeException("\"metadata\" property is reserved");
-            }
-
-            validatePartitionKeys(eventType);
 
-            SchemaLoader.load(schemaAsJson);
-        } catch (final JSONException e) {
-            throw new InvalidEventTypeException("schema must be a valid json");
-        } catch (final SchemaException e) {
-            throw new InvalidEventTypeException("schema must be a valid json-schema");
+        final List<String> absentFields = eventType.getPartitionKeyFields().stream()
+                .filter(field -> !schema.definesProperty(convertToJSONPointer(field)))
+                .collect(Collectors.toList());
+        if (!absentFields.isEmpty()) {
+            throw new InvalidEventTypeException("partition_key_fields " + absentFields + " absent in schema");
         }
     }
 
     private void assignTopic(final EventType eventType) {
         eventType.setTopic(uuidGenerator.randomUUID().toString());
     }
 
-    private boolean hasReservedField(final JSONObject schemaAsJson, final String field) {
-        return schemaAsJson.optJSONObject("properties") != null
-                && schemaAsJson.getJSONObject("properties").has(field);
+    private String convertToJSONPointer(final String value) {
+        return value.replaceAll("\\.", "/");
```

Comment:
I would not like to go into the details of JSON Pointer, since our previous path representation is simplified, but maybe we should have another ticket to avoid definition of properties with weird string, like https://tools.ietf.org/html/rfc6901#section-5

Because if we allow that, then this transformation could fail.